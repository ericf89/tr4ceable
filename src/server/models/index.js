import Router from 'express-promise-router';
import userRoutes from './user/routes';
import authRoutes from './auth/routes';

const router = Router();

router.use('/auth', authRoutes);
router.use('/users', userRoutes);

export default router;

