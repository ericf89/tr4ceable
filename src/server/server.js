import express from 'express';
import bodyParser from 'body-parser';
import config from 'config';
import compression from 'compression';
import morgan from 'morgan';
import winston from 'winston';
import mongoose from 'mongoose';
import apiRoutes from './models';

mongoose.Promise = Promise;

export const mongoConnect = async () => {
  await mongoose.connect(config.get('mongoUrl'));
  return async () => {
    if (config.get('nodeEnv') === 'test') {
      await mongoose.connection.dropDatabase();
    }
    return mongoose.disconnect();
  };
};

export const server = () => {
  const app = express();

  app.set('port', config.get('port'));
  app.disable('x-powered-by');
  app.use(bodyParser.json());
  app.use(compression());

  app.use('/api', apiRoutes);
  app.get('*', (req, res) => res.sendStatus(404));
  return app;
};

export default (openDbConnection = mongoConnect, serverConfig = server) => {
  openDbConnection();
  const app = serverConfig();
  const development = config.get('nodeEnv') === 'development';
  app.locals.pretty = development;
  app.use(morgan(development ? 'dev' : 'combined'));
  app.listen(config.get('port'), () => winston.info('Listening on port %d', config.get('port')));
};

