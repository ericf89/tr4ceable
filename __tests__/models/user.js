import supertest from 'supertest';
import { server, mongoConnect } from '../../src/server/server';

import User from '../../src/server/models/user/db';


describe('Users', () => {
  let disconnect;
  beforeAll(async () => (disconnect = await mongoConnect(__filename)));
  beforeEach(async () => { await User.find().remove(); });
  afterAll(() => disconnect());

  const mockUserBody = {
    email: 'eric@eric.com',
    password: 'hunter1',
  };

  it('should 201 when POSTing to create', async () => {
    const { body, statusCode } = await supertest(server())
      .post('/api/users')
      .send(mockUserBody);

    expect(body._id);
    expect(body.email).toEqual(mockUserBody.email);
    expect(body.admin).toBeFalsy();
    expect(body.createdAt).toBeTruthy();
    expect(body.updatedAt).toBeTruthy();
    expect(statusCode).toBe(201);
  });

  it('should 400 if you try to create a user with a taken email', async () => {
    await supertest(server())
      .post('/api/users')
      .send(mockUserBody);

    const { body, statusCode } = await supertest(server())
      .post('/api/users')
      .send(mockUserBody);

    // Check that id exists, and then delete it before testing snapshot,
    // since it'll always be unique.
    expect(statusCode).toBe(400);
    expect(body).toMatchSnapshot();
  });

  it('should 200 when GETing by id', async () => {
    const testUser = await User.create(mockUserBody);

    const { body, statusCode } = await supertest(server())
      .get(`/api/users/${testUser.id}`);

    expect(statusCode).toBe(200);
    expect(body._id).toEqual(testUser.id);
    expect(body.email).toEqual(mockUserBody.email);
    expect(body.admin).toBeFalsy();
    expect(body.createdAt).toBeTruthy();
    expect(body.updatedAt).toBeTruthy();
  });

  it('should 200 when GETing by email', async () => {
    const testUser = await User.create(mockUserBody);
    const { body, statusCode } = await supertest(server())
      .get(`/api/users/?email=${mockUserBody.email}`);

    expect(statusCode).toBe(200);
    expect(body.users.length).toBe(1);
    const { users: [user] } = body;
    expect(user._id).toEqual(testUser.id);
    expect(user.email).toEqual(mockUserBody.email);
    expect(user.admin).toBeFalsy();
    expect(user.createdAt).toBeTruthy();
    expect(user.updatedAt).toBeTruthy();
  });

  it('should 404 when GETing by email if the email hasn\'t registered', async () => {
    const { body, statusCode } = await supertest(server())
      .get('/api/users/?email=nope@nope.com');

    expect(statusCode).toBe(404);
    expect(body).toMatchSnapshot();
  });
});
