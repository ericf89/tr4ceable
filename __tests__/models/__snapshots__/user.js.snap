// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users should 400 if you try to create a user with a taken email 1`] = `
Object {
  "error": "E11000 duplicate key error index: tr4ce_TEST.users.$email_1 dup key: { : \\"eric@eric.com\\" }",
  "message": "Invalid data for user",
}
`;

exports[`Users should 404 when GETing by email if the email hasn't registered 1`] = `
Object {
  "users": Array [],
}
`;
